server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -478,19 +478,23 @@ public Translog() {}
 
         public Translog(StreamInput in) throws IOException {
             super(in);
-            recovered = in.readVInt();
-            total = in.readVInt();
-            totalOnStart = in.readVInt();
-            totalLocal = in.readVInt();
+            synchronized (this) {
+                recovered = in.readVInt();
+                total = in.readVInt();
+                totalOnStart = in.readVInt();
+                totalLocal = in.readVInt();
+            }
```

Comment:
Looks like this is a constructor. Please correct me if I am wrong, but I do not think it is possible to modify object from other thread while it is still constructed.

Reply:
++ unless the constructor were to spawn a thread this is impossible and there's no reason to at sync here

Reply:
That's a good point! Let me remove it.

I guess it'd nice to have some jcstress tests here in order to check that the synchronization is correct.

Reply:
Actually, it looks like we still need to have a synchornized block here in order to guarantee visibility. The `recovered`, `total`, `totalOnStart`, `totalLocal` fields are not finals, so we don't have a write fence to guarantee that they will be correctly published for a different thread.

Reply:
Background info for that: https://shipilev.net/blog/2014/safe-public-construction/#_safe_initialization

Reply:
:+1: Then I would only suggest to put a comment in constructor explaining why synchronized is needed there

Reply:
The issue is with the visibility of the constructor. The compiler doesn't  a write barrier, hence there no is happens-before relationship between writing to to total* fields and reading from them.

Reply:
It seems like as long as the object is safely published, subsequent synchronized reads should get full visibility of the object fields (see https://stackoverflow.com/a/39393985) But maybe I'm missing something here.

Reply:
But that's the crux of the issue!
```
        private int recovered;
        private int total = UNKNOWN;
        private int totalOnStart = UNKNOWN;
        private int totalLocal = UNKNOWN;
```
The fields are neither volatile nor final, so the object is unsafely published.